ILP placer for medium sized blocks

srini229 opened the pull request:
ILP formulation:
![ilp_place-2](https://user-images.githubusercontent.com/73570678/155577124-99513b70-7d2f-4d7f-9989-3d88d5d8b819.png)
![ilp_place-3](https://user-images.githubusercontent.com/73570678/155577146-ab648d99-1f65-4300-902a-1c669aeb5e65.png)

srini229 commented:
Comments from @soneryaldiz : 
By the way, the existing tests with existing placer are failing on this branch. This would make it hard to create benchmarking tests. 
FAILED test_circuits.py::test_ota_six - AssertionError: ckt_ota_six: No results generated
FAILED test_circuits.py::test_ro_simple - AssertionError: ckt_ro_simple: No results generated
FAILED test_circuits.py::test_common_source - AssertionError: ckt_common_source: No results generated

soneryaldiz commented:
@srini229 
I started `tests/pdk/finfet_pdk/test_benchmark_place.py` to compare runtime/area/hpwl. However, the default placer is not working on this branch.

srini229 commented:
@soneryaldiz select_in_ILP was enabling in the earlier implementation of selection from nvariants used in SA placer. I commented it temporarily in this branch. We can decide whether to retain/remove it in a separate PR.

test_benchmark_place.py passes now.

soneryaldiz commented:
@srini229 . I added a second benchmark that is failing.

srini229 commented:
@soneryaldiz my mistake on an assumption that two ordering constraints won't have one or more blocks in common. fixed it now. test_benchmark_place is passing now.

srini229 commented:
@soneryaldiz aspect ratio check was missing in the ilp placer and i assumed there wouldn't be any shapes in common between ordering constraints. fixed both now. test_benchmark_place is passing now.

soneryaldiz commented:
@srini229 , below is a comparison over 3 cases. Only for the third case, there is a noticeable increase in runtime (15 sec vs 100sec).
```
ckt_b1_tamu_sp: AREA=23.13 HPWL=44.81 TIME=3.17
ckt_b1_umn_ilp: AREA=23.13 HPWL=44.81 TIME=2.93
ckt_b2_tamu_sp: AREA=47.63 HPWL=87.53 TIME=15.77
ckt_b2_umn_ilp: AREA=47.63 HPWL=87.53 TIME=15.65
ckt_b3_tamu_sp: AREA=34.29 HPWL=75.70 TIME=15.48
ckt_b3_umn_ilp: AREA=34.29 HPWL=75.70 TIME=101.98
``` 
To get the numbers above: Set BENCHMARK to True in test_benchmark_place.py and run `pytest -vv -s test_benchmark_place.py` 
I am skipping these tests from CI.

srini229 commented:
@soneryaldiz I found a bug in the same-template constraint handling in the ILP placer. With the fix, B2 has better HPWL with ILP. For B3, as expected the runtime is significantly larger. This is because of size of the problem (~10 blocks), which approaches the limit of ILP to run in reasonable time.
```
ckt_b1_tamu_sp: AREA=23.13 HPWL=44.81 TIME=4.23
ckt_b1_umn_ilp: AREA=23.13 HPWL=44.81 TIME=3.46
ckt_b2_tamu_sp: AREA=47.63 HPWL=87.53 TIME=54.92
ckt_b2_umn_ilp: AREA=47.63 HPWL=85.71 TIME=13.35
ckt_b3_tamu_sp: AREA=34.29 HPWL=75.70 TIME=56.52
ckt_b3_umn_ilp: AREA=34.29 HPWL=75.70 TIME=148.94
```
The ILP solver used in this branch is multi-threaded. I tried running it with 8 threads and got: 
```ckt_b3_umn_ilp: AREA=34.29 HPWL=75.70 TIME=30.39```
This can be enabled using the NUM_THREADS hyper param in align/pnr/toplevel.py.

soneryaldiz commented:
@srini229 , thanks. By the way, b2 and b3 are the comparator circuit with different set of constraints. It appears that just changing the constraints could also have quite an impact on runtime.

srini229 commented:
@soneryaldiz yes. the number of constraints determine the number of 'binary' variables used to determine overlap between pairs of blocks. The two images in https://github.com/ALIGN-analoglayout/ALIGN-public/pull/989#issue-1149561736 illustrates this. Two binary variables are required to define no-overlap between a pair of blocks. For N binary variables, the search space is 2^N. This formulation can be intermediate between the enumeration placer and the SA placer where there are many constraints or if the block size is slightly larger than what enum. placer can handle.

soneryaldiz commented:
@srini229 , is there any updates on this PR?

soneryaldiz commented:
More specifically on handling PlaceOnGrid constraints?

srini229 commented:
@soneryaldiz this should already handle PlaceOnGrid constraint. The pending task is to make this as an intermediate placer in between the enumerative placer and the SA placer based on the constraints.

srini229 commented:
@soneryaldiz I added a hyper param to enable/disable the ILP placer.
It can be enabled using `hyper.use_ILP_placer` in `align/placer.py`.
I changed it from the select_in_ILP switch since this switch is being used currently for selecting variants in the SA placer ILP.

There was a bug in the placeongrid constraint handling in the ILP placer. 
I have fixed it now and the test `test_placement_grid.py::test_cmp_on_grid` passes.

srini229 commented:
@soneryaldiz this branch is ready to merge. You can enable the ILP placer using the hyper param: `hyper.use_ILP_placer`.

soneryaldiz commented on PlaceRouteHierFlow/PnR-pybind11.cpp:
```diff
@@ -492,7 +492,9 @@ PYBIND11_MODULE(PnR, m) {
     .def_readwrite("max_init_trial_count", &PlacerHyperparameters::max_init_trial_count)
     .def_readwrite("max_cache_hit_count", &PlacerHyperparameters::max_cache_hit_count)
     .def_readwrite("select_in_ILP", &PlacerHyperparameters::select_in_ILP)
+    .def_readwrite("use_ILP_placer", &PlacerHyperparameters::use_ILP_placer)
```
It would be nice to add comments to describe what select_in_ILP, use_ILP_placer and ilp_solver do.

soneryaldiz commented on align/gdsconv/gen_pl_gds.py:
```diff
@@ -0,0 +1,127 @@
+#!/usr/bin/env python
+
```
How is this related to ILP placer?

srini229 replied:
It is not relevant. Just a script to visualize the placement as GDS. I'll remove it since a better version is already in master now.

soneryaldiz replied:
If you still want to keep this, you can move it to dev/Experimental area so that it does not impact coverage.

soneryaldiz commented on align/schema/checker.py:
```diff
@@ -182,7 +182,7 @@ def solve(self):
         if r == z3.unsat:
             z3.set_option(max_depth=10000, max_args=100, max_lines=10000)
-            logger.debug(f"Unsat encountered: {self._solver}")
+            logger.info(f"Unsat encountered: {self._solver}")
```
Is there a particular reason to make this info? This may reveal compiler's attempts to add constraints and confuse user.

soneryaldiz commented on tests/pdk/finfet_pdk/test_benchmark_place.py:
```diff
@@ -0,0 +1,134 @@
+# ]
+PARAMS = [
+        ("tamu_sp", [""]),
+        ("umn_ilp", ["--select_in_ILP"])
```
is this test configuration no longer correct?

srini229 replied:
Fixed it in the recent checkin.

soneryaldiz commented on tests/test_cmdline.py:
```diff
@@ -48,8 +48,6 @@ def test_cmdline(design,num_placements,PDN_mode,select_in_ILP):
     if PDN_mode:
         args.append( '--PDN_mode')
-    if select_in_ILP:
-        args.append( '--select_in_ILP')
```
Why this removal?

srini229 replied:
My mistake. All these removal for debugging. I'm reverting these.

soneryaldiz commented:
@srini229 , it would be good to make sure the added code has coverage with small circuit examples.

soneryaldiz commented on PlaceRouteHierFlow/PnR-pybind11.cpp:
```diff
@@ -491,8 +491,10 @@ PYBIND11_MODULE(PnR, m) {
     .def_readwrite("max_cache_hit_count", &PlacerHyperparameters::max_cache_hit_count)
-    .def_readwrite("select_in_ILP", &PlacerHyperparameters::select_in_ILP)
-    .def_readwrite("ilp_solver", &PlacerHyperparameters::ilp_solver)
+    .def_readwrite("select_in_ILP", &PlacerHyperparameters::select_in_ILP) // select variant of blocks in SA using ILP used for compaction
```
Does this mean?
select_in_ILP == False => Variants will be chosen in sequence pair representation
select_in_ILP == True => Variants will be chosen by ILP solver?

srini229 replied:
@soneryaldiz Yes. This was added by @854768750 upon Steve's proposal sometime ago.

854768750 replied:
The old version with lpsolve uses this.

soneryaldiz replied:
@854768750 , could you please elaborate? What do you mean by old version? Is this switch no longer functional?

stevenmburns replied:
We used to have tests for this but it was so slow we commented it out.

854768750 replied:
The number of decision variables may be several times more if variants a are chosen by ILP solver

srini229 commented:
> @srini229 , it would be good to make sure the added code has coverage with small circuit examples.

@soneryaldiz I enabled couple of tests in the benchmark you had added. This should cover most of the scenarios in the added code. I also added a copy of test_cmp_on_grid to run using ILP placer so that the place on grid portion is tested as well.

soneryaldiz reviewed (approved):
Many thanks for doing this!

soneryaldiz commented:
@srini229 , looking at the two slides on top, does not align constraints also simplify the formulation similar to symmetry?

srini229 commented:
> @srini229 , looking at the two slides on top, does not align constraints also simplify the formulation similar to symmetry?

@soneryaldiz yes it does. Similar is the case with order as well. These are in the code. I can make an elaborate version of the slides and upload them to the ALIGN google docs.

soneryaldiz commented:
Thanks @srini229
